Source/Core/Core/IOS/DI/DI.cpp

```diff
@@ -30,93 +34,617 @@ void DI::DoState(PointerWrap& p)
+{
+  if (!m_has_initialized)
+  {
+    ResetDIRegisters();
+    m_has_initialized = true;
+  }
+  return Device::Open(request);
 }
 
 IPCCommandResult DI::IOCtl(const IOCtlRequest& request)
 {
+  if (!m_has_initialized)
+  {
+    ResetDIRegisters();
+    m_has_initialized = true;
+  }
+
   // DI IOCtls are handled in a special way by Dolphin
   // compared to other IOS functions.
   // This is a wrapper around DVDInterface's ExecuteCommand,
   // which will execute commands more or less asynchronously.
   // Only one command can be executed at a time, so commands
   // are queued until DVDInterface is ready to handle them.
 
+  u8 command = Memory::Read_U8(request.buffer_in);
+  if (request.request != command)
+  {
+    WARN_LOG(
+        IOS_DI,
+        "IOCtl: Received conflicting commands: ioctl 0x%02x, buffer 0x%02x.  Using ioctl command.",
+        request.request, command);
+  }
+
+  // Some commands don't interact directly with DVDInterface.  Handle those here:
+  // TODO: However, this can lead to a command replying before one of the DVDInterface commands is
+  // done, which is inaccurate.
   bool ready_to_execute = m_commands_to_execute.empty();
   m_commands_to_execute.push_back(request.address);
+
   if (ready_to_execute)
-    StartIOCtl(request);
+  {
+    ProcessQueuedIOCtl();
+  }
 
   // DVDInterface handles the timing and will call FinishIOCtl after the command
   // has been executed to reply to the request, so we shouldn't reply here.
   return GetNoReply();
 }
 
-void DI::StartIOCtl(const IOCtlRequest& request)
+void DI::ProcessQueuedIOCtl()
+{
+  if (m_commands_to_execute.empty())
+  {
+    PanicAlert("IOS::HLE::Device::DI: There is no command to execute!");
+    return;
+  }
+
+  IOCtlRequest request{m_commands_to_execute.front()};
+
+  auto finished = StartIOCtl(request);
+  if (finished)
+  {
+    CoreTiming::ScheduleEvent(2700 * SystemTimers::TIMER_RATIO, s_finish_executing_di_command,
+                              static_cast<u64>(finished.value()));
+    return;
+  }
+}
+
+std::optional<DI::DIResult> DI::WriteIfFits(const IOCtlRequest& request, u32 value)
+{
+  if (request.buffer_out_size < 4)
+  {
+    WARN_LOG(IOS_DI, "Output buffer is too small to contain result; returning security error");
+    return std::make_optional(DIResult::SECURITY_ERROR);
+  }
+  else
+  {
+    Memory::Write_U32(value, request.buffer_out);
+    return std::make_optional(DIResult::SUCCESS);
+  }
+}
+
+std::optional<DI::DIResult> DI::StartIOCtl(const IOCtlRequest& request)
 {
-  const u32 command_0 = Memory::Read_U32(request.buffer_in);
-  const u32 command_1 = Memory::Read_U32(request.buffer_in + 4);
-  const u32 command_2 = Memory::Read_U32(request.buffer_in + 8);
+  if (request.buffer_in_size != 0x20)
+  {
+    ERROR_LOG(IOS_DI, "IOCtl: Received bad input buffer size 0x%02x, should be 0x20",
+              request.buffer_in_size);
+    return std::make_optional(DIResult::SECURITY_ERROR);
```

Comment:
As far as I'm aware, you can simply return the enum value directly without the need for `std::make_optional`, Ditto elsewhere where applicable.

Reply:
Visual studio didn't seem to like that before, but it looks like it works now.  Possibly it didn't work before due to me using a regular `enum` instead of an `enum class`?